ksql-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java

```diff
@@ -227,12 +229,55 @@ public SchemaKStream into(
           processingLogger);
     }
 
-    private Optional<Field> findKeyField(final List<SelectExpression> selectExpressions) {
-      if (!getKeyField().isPresent()) {
+    private KeyField findKeyField(final List<SelectExpression> selectExpressions) {
+      return KeyField.of(findNewKeyField(selectExpressions), findLegacyKeyField(selectExpressions));
+    }
+
+    private Optional<String> findNewKeyField(final List<SelectExpression> selectExpressions) {
```

Comment:
Can we make this and `findLegacyKeyField` the same function that takes a `Supplier<String> keyFieldSupplier` to get the field name?

Reply:
Nope, the logic is different. I did take a stab at it but the code ends up very ugly.